mother/src/drive.c

```diff
@@ -0,0 +1,258 @@
+	if (drive->feedback_callback(feedback, feedback_buffer_size,
+				     drive->config.wheels_per_side)) {
+		// ERROR: Something went wrong while getting feedback
+	}
+	for (int i = 0; i < drive->config.wheels_per_side; i++) {
+		const float left_feedback = feedback[i];
+		const float right_feedback = feedback[drive->config.wheels_per_side + i];
+
+		if (isnan(left_feedback) || isnan(right_feedback)) {
+			// ERROR: One of the wheels gives invalid feedback
+		}
+
+		left_feedback_mean += left_feedback;
+		right_feedback_mean += right_feedback;
+	}
+	k_free(feedback);
+	left_feedback_mean /= drive->config.wheels_per_side;
+	right_feedback_mean /= drive->config.wheels_per_side;
+
+	if (drive->config.update_type | POSITION_FEEDBACK) {
+		// TODO: odometry update from position
+	} else {
+		diffdrive_odometry_update_from_velocity(
+			drive->odometry,
+			left_feedback_mean * left_wheel_radius * 1.0f,   // FIXME period.seconds()
+			right_feedback_mean * right_wheel_radius * 1.0f, // FIXME period.second()
+			now_timestamp);
+	}
+
+	const float velocity_left =
+		(linear_command - angular_command * wheel_separation / 2.0) / left_wheel_radius;
+	const float velocity_right =
+		(linear_command + angular_command * wheel_separation / 2.0) / right_wheel_radius;
+
+	float *velocity_buffer = (float *)k_malloc(sizeof(float) * feedback_buffer_size);
+	for (int i = 0; i < drive->config.wheels_per_side; i++) {
+		velocity_buffer[i] = velocity_left;
+		velocity_buffer[drive->config.wheels_per_side + i] = velocity_right;
+	}
+	if (drive->velocity_callback(velocity_buffer, feedback_buffer_size,
+				     drive->config.wheels_per_side)) {
+		// ERROR: Something went wrong writing the velocities
+	}
+	k_free(velocity_buffer);
+}
+
+struct FloatRollingMeanAccumulator {
+	float *buffer;
+	const int buffer_len;
+	int next_insert;
+	float sum;
+	int buffer_filled;
+};
+
+void *float_rolling_mean_accumulator_init(int rolling_window_size)
+{
+	float *buffer = k_calloc(rolling_window_size, sizeof(float)); 
+	struct FloatRollingMeanAccumulator frma = {.buffer = buffer,
+						   .buffer_len = rolling_window_size,
+						   .buffer_filled = false,
+						   .next_insert = 0,
+						   .sum = 0.0f};
+	void *heap_frma = k_malloc(sizeof(frma));
+	// FIXME memcpy(heap_frma, frma); AMB &frma
+	memcpy(heap_frma, &frma, sizeof(frma));
+	return heap_frma;
+}
+
+void float_rolling_mean_accumulator_accumulate(struct FloatRollingMeanAccumulator *frma, float val)
+{
+	frma->sum -= frma->buffer[frma->next_insert];
+	frma->sum += val;
+	frma->buffer[frma->next_insert] = val;
+	frma->next_insert++;
+	frma->buffer_filled |= frma->next_insert >= frma->buffer_len;
+	frma->next_insert = frma->next_insert % frma->buffer_len;
+}
+
+float float_rolling_mean_accumulator_get_rolling_mean(
+	const struct FloatRollingMeanAccumulator *frma)
+{
+	int valid_data_count =
+		frma->buffer_filled * frma->buffer_len + !frma->buffer_filled * frma->next_insert;
+	// CHECKME: assert(valid_data_count > 0);
```

Comment:
you need some kind assert func here or just if else is suffice

Reply:
What do you recommend? How costly is a runtime assert in Zephyr? Does it reset or panic the µC?